### PR TITLE
Implement recommended extension/server setup

### PR DESCRIPTION
**Pull request by WilsonZiweiWang**
### Why
I want to get rid of this message that showed up during the execution of the script `installServerIntoExtension`:
> The script is deprecated. See https://github.com/microsoft/vscode-extension-samples/tree/master/lsp-sample for an example on how to setup an extension / server project.

Source of the deprecation message: 
https://github.com/microsoft/vscode-languageserver-node/blob/0226f6cd925d279dda2b777d25c5b69518ac0035/server/bin/installServerIntoExtension#L70

Ticket: 14334

### How
To implement the recommended setup in `lsp-sample`:
1. The extension manifest needs to be at the workspace root's package.json. (`vsce package` will need to run in root afterwards)
2. Files in the server and the client should be compiled to their own `out/`

Other changes to make:
1. Need to update the root's README for it will be used as the extension information in the marketplace.
2. Update paths in `launch.json`, `main.yml`.
3. Update ignore patterns in `.gitignore`, `.vscodeignore` and `.npmignore`.

### Test results
All functionalities seem to work on my end
**VSIX**
image

**npm package**
image

### What is not working
~The integration tests don't seem to work. Local run and CI froze at this stage.~

### CODE REVIEW

**Comment by deribaucourt** [comment]
Probably resolves #178
Very excited to get this. All packages, debug configurations and test configurations must be fully working though.

**Comment by WilsonZiweiWang** [comment]
The client's README content needs to move to the root. Need to decide what to do with the existing one.

File: .vscode/launch.json
Comment (deribaucourt) [reply .vscode/launch.json:74]: Other debug configurations run well :+1:

File: .github/workflows/main.yml
Comment (deribaucourt) [inline .github/workflows/main.yml:96]: It used to be called by npm pack the line bellow.
I think the change is good, but could you just explain it to me briefly?

File: .vscodeignore
Comment (deribaucourt) [inline .vscodeignore:1]: gotta check that we don't package unwanted files

File: images/bitbake-terminal-profile.png
Comment (deribaucourt) [inline images/bitbake-terminal-profile.png:1]: I have to check if the marketplace README properly points to the new URLs once we have pushed to main.

File: .vscode/launch.json
Comment (deribaucourt) [inline .vscode/launch.json:45]: We must check that any code modification in client and server is properly recompiled before all debug launch configurations.

Comment (deribaucourt) [reply .vscode/launch.json:45]: `npm run compile` at the root used to recompile both client and server. Is it still the case?

Comment (deribaucourt) [reply .vscode/launch.json:45]: It looks like it's still the case, but much faster! great job!

File: .vscodeignore
Comment (deribaucourt) [inline .vscodeignore:16]: gitignore needs some updates.:
 - `/*.vsix`

File: server/package.json
Comment (deribaucourt) [inline server/package.json:29]: Getting command not found when starting language-server-bitbake for vim
image
To reproduce, 
 - cd server
 - npm pack
 - npm install -g ./*.gz
 - language-server-bitbake --stdio (should start and await input)

File: .vscodeignore
Comment (deribaucourt) [reply .vscodeignore:1]: looks appropriate in `~/.vscode/extensions/yocto-project.yocto-bitbake-2.4.0`. Could you also take a look and see if there's anything that should be removed? Maybe server node_modules?

File: server/package.json
Comment (deribaucourt) [reply server/package.json:29]: I think once again you forgot to update the server's package.json bin location (out/server.js instead of out/src/server.js)

Comment (deribaucourt) [reply server/package.json:29]: Make sure to also test that the resources paths are all correct (docs and .wasm) once packaged for vim and VSCode.

File: images/bitbake-terminal-profile.png
Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: First, they are broken once the .vsix is installed in the extension view
image

Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: I think you need to update the images URLs in the root README.md. Check if any other resource needs updated URLs.
I know the Marketplace needs different tricks to also render them online. (they are not installed, it tries to get them from github I think). I can't remember how to test this. Could you take a look at this and make sure the README renders properly in:
 - debug launch configuration
 - installed through .vsix
 - online marketplace page

**Review by deribaucourt — CHANGES_REQUESTED** [review]

File: .vscode/launch.json
Comment (WilsonZiweiWang) [reply .vscode/launch.json:74]: Added outfiles from integration-tests

File: .vscodeignore
Comment (WilsonZiweiWang) [reply .vscodeignore:16]: Done

File: server/package.json
Comment (WilsonZiweiWang) [reply server/package.json:29]: Yes, my bad. `bin` updated.

File: images/bitbake-terminal-profile.png
Comment (WilsonZiweiWang) [reply images/bitbake-terminal-profile.png:1]: I already updated the URLs, they work on the github page. https://github.com/savoirfairelinux/vscode-bitbake/tree/use-recommended-extension-server-structure

How did you get the extension page with the broken image? The images are showing on my end. (After installing the vsix)

File: .vscodeignore
Comment (deribaucourt) [reply .vscodeignore:1]: Perfect, no need to address package size.

File: .vscode/launch.json
Comment (deribaucourt) [reply .vscode/launch.json:74]: Works now

File: server/package.json
Comment (deribaucourt) [reply server/package.json:29]: Now working on NVIM

File: images/bitbake-terminal-profile.png
Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: I packaged the .vsix and installed it. If I have a look at `~/.vscode/extensions/yocto-project.yocto-bitbake-2.4.0/README.md`
It points to images in 
```
![BitBake terminal profile](https://github.com/yoctoproject/vscode-bitbake/raw/HEAD/client/images/bitbake-terminal-profile.png)
```
(see how vsce converts README images for online marketplace display)
I think it's correct. This will be where they'll be located once this PR is merged?

File: .vscodeignore
Comment (deribaucourt) [reply .vscodeignore:9]: image

File: images/bitbake-terminal-profile.png
Comment (WilsonZiweiWang) [reply images/bitbake-terminal-profile.png:1]: It shouldn't point to the `client/images`, so the current README won't work. One way to fix it is to remove the `vsce` property in the package.json. `vsce` will now by default use `blob/HEAD/`. Or we can just remove the `client` at the end.
image

File: .vscodeignore
Comment (WilsonZiweiWang) [reply .vscodeignore:9]: something like:
```
"dev:ci-vsix": "npm run clean && npm install ..."
"dev:ci-npm": "npm run clean && npm install ..." 
```

File: images/bitbake-terminal-profile.png
Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: If merging this would break the current README, we can set the default branch to "main" temporarily. that way the URL will point to the previous version.

File: .vscodeignore
Comment (deribaucourt) [reply .vscodeignore:9]: I often need to package locally, in which case I may have .tgz lying around. the ignore file should be updated.

File: images/bitbake-terminal-profile.png
Comment (WilsonZiweiWang) [reply images/bitbake-terminal-profile.png:1]: With e9b41668cb3c5a785adc6e765a559486ec0348b5 the URLs should be correct.
```
Instructions to build, test and debug the extension are available in the [README-DEVELOPER](https://github.com/yoctoproject/vscode-bitbake/raw/HEAD/README-DEVELOPER.md).
```

File: .vscodeignore
Comment (WilsonZiweiWang) [reply .vscodeignore:9]: I package locally a lot too in this PR, that's exactly why I proposed the scripts above which follow the same steps in the CI such that we have the most accurate result. In general, the ignore patterns should be relevant to everyone on the team. But adding patterns like .tgz and .vsix can help prevent mistakes in the development, so I will add them.

**Review by deribaucourt — COMMENTED** [review]
Code LGTM, discussions

File: images/bitbake-terminal-profile.png
Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: Will probably work, but if we merge this, we break the current version's marketplace page right?
I think we have to put main as the default repository branch with the old version untill next release?

Comment (WilsonZiweiWang) [reply images/bitbake-terminal-profile.png:1]: Not sure I follow. In this PR, we need to run `vsce package` in the project root (not in `/client/` anymore), because the extension manifest is now the project root's `package.json`. Therefore, the project root's README is used for the marketplace page. And now we moved everything from the previous client's README to the project root's README, so everything should be fine when this PR is merged.

Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: The currently published 2.3 version will keep the old README which uses images URLs from the old README.

Comment (WilsonZiweiWang) [reply images/bitbake-terminal-profile.png:1]: Oh, I see. You meant the default repo branch on Github, not something we can specify in the package.json. I think we could make `main` as the default for now as you said or we just wait till the release to merge it.

Comment (deribaucourt) [reply images/bitbake-terminal-profile.png:1]: I want to merge this to reduce git conflicts as well as have it around longer for testing. Changing the default branch to main and approving the PR.

**Review by deribaucourt — APPROVED** [review]

**Pull request merged**
